dali/python/nvidia/dali/external_source.py

```diff
@@ -0,0 +1,258 @@
+# custom wrappers around ops
+from future.utils import with_metaclass
+from nvidia.dali.ops import _DaliOperatorMeta, _OperatorInstance, _gpu_ops, _cpu_ops
+from nvidia.dali import backend as _b
+import itertools
+
+def _check_data_batch(data, batch_size, layout):
+    if isinstance(data, (list, tuple)):
+        if len(data) != batch_size:
+            raise RuntimeError("The external source callback returned an unexpected batch "
+            "size: {} instead of {}".format(len(data), batch_size))
+        if len(data) > 0:
+            dim = len(data[0].shape)
+            for t in data:
+                if len(t.shape) != dim:
+                    raise RuntimeErrorError("All tensors in a batch must have the same number of dimensions")
+            if layout != "" and dim != len(layout):
+                raise RuntimeError("The layout '{}' cannot describe {}-dimensional data".format(layout, dim))
+    else:
+        dim = len(data.shape) - 1
+        if data.shape[0] != batch_size:
+            raise RuntimeError("The external source returned an unexpected batch "
+            "size: {} instead of {}".format(data.shape[0], batch_size))
+        if layout != "" and dim != len(layout):
+            raise RuntimeError("The layout '{}' cannot describe {}-dimensional data".format(layout, dim))
+
+
+class _ExternalSourceGroup(object):
+    def __init__(self, callback, is_multioutput, instances = []):
+        self.instances = list(instances)  # we need a copy!
+        self.is_multioutput = is_multioutput
+        self.callback = callback
+        if callback is not None:
+            if callback.__code__.co_argcount not in [0, 1]:
+                raise TypeError("External source callback must be a callable with 0 or 1 argument")
+            self.accepts_iter_num = (callback.__code__.co_argcount == 1)
+        else:
+            self.accepts_iter_num = None
+
+    def append(self, instance):
+        self.instances.append(instance)
+
+    def call_and_feed(self, pipeline, current_iter):
+        if self.accepts_iter_num:
+            callback_out = self.callback(current_iter)
+        else:
+            callback_out = self.callback()
+
+        if self.is_multioutput:
```

Comment:
Why we need `is_multioutput`? Can't it just check the `len(self.instances)`?

Reply:
No; when using num_outputs != None, the the source is expected to return a list (or tuple) of batches, even if num_outputs == 1.

Reply:
Can we ask the source to always return an iterable even if num_outputs == None?

Reply:
On the other hand, where there is only one output it would pose an unnecessary restriction to the user.